File: 백준/Gold/16234. 인구 이동/인구 이동.py
import sys
from collections import deque
input = sys.stdin.readline

# N, L, R
# N 만큼 반복
# L명 이상 R명 이하 -> 하루동안 문 염
# 국경선 다 열리면 인구이동
# 인접한 카만 이동 가능, 그 나라는 하루동안 연합
# 연합을 이루고 있는 각 칸의 인구수는 (연합의 인구수) / (연합을 이루고 있는 칸의 개수)가 된다. 편의상 소수점은 버린다.
# 연합을 해체하고, 모든 국경선을 닫는다.

# 3 5 10
# 10 15 20
# 20 30 25
# 40 22 10

# 5명이상 10명 이하
# 10 : 15 : 20
# . .  --   . .
# 20 : 30 : 25    -> 10+ 15+ 20+ 20+ 30+ 25 = 120 // 6 == 20 
# -------------
# 40 | 22 | 10    -> 40 + 22 + 10 = 72//3 = 24

# 20 | 20 | 20
# ------------
# 20 | 20 | 20
# ------------
# 24 | 24 | 24  -- > 20 + 20 + 20+ 24 + 24+ 24 == 132 // 6 = 22

N,L,R = map(int, input().split())
population = []
dx = [-1, 0, 1, 0]
dy = [0, 1, 0 ,-1]
pos = []
for _ in range(N):
    population.append(list(map(int, input().split())))
day = 0
def bfs(x,y, move, area, visited):
    global pos, day
    visited[x][y] = True
    q = deque()
    q.append([x,y])
    pos.append([x,y])
    while q:
        x, y = q.popleft()
        for i in range(4):
            nextX = dx[i] + x
            nextY = dy[i] + y
            if 0 <= nextX < N and 0 <= nextY < N and not visited[nextX][nextY]:
                if L <= abs(population[nextX][nextY] - population[x][y]) <= R:
                    move += population[nextX][nextY]
                    visited[nextX][nextY] = True
                    area += 1
                    q.append([nextX, nextY])
                    pos.append([nextX,nextY])
    return move, area
area = 0
while True:
    day += 1
    flag = False
    visited = [[False for _ in range(N)] for _ in range(N)]
    for i in range(N):
        for j in range(N):
            if not visited[i][j]:
                move, area = bfs(i, j, population[i][j], 1, visited)
                if len(pos) >= 2:
                    flag = True
                    person = move // area
                    for x,y in pos:
                        population[x][y] = person
                pos = []
    if not flag:
        break
print(day-1)

File: 백준/Gold/16234. 인구 이동/인구 이동.java
import java.io.*;
import java.util.*;

public class Main {
	/*
2 20 50
50 30
20 40
	 */

	static int[] dx = {-1, 0, 1, 0};
	static int[] dy = {0, 1, 0, -1};
	static boolean[][] visited;
	static int[][] country;
	static ArrayList<int[]> moveInfo;

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		StringTokenizer st;

		st = new StringTokenizer(br.readLine());
		int N = Integer.parseInt(st.nextToken());
		int L = Integer.parseInt(st.nextToken());
		int R = Integer.parseInt(st.nextToken());

		country = new int[N][N];

		for (int i = 0; i < N; i++) {
			st = new StringTokenizer(br.readLine());

			for (int j = 0; j < N; j++) {
				country[i][j] = Integer.parseInt(st.nextToken());
			}
		}

		int answer = solve(N, L, R);
		System.out.println(answer);
	}

	private static int solve(int N, int L, int R) {

		int day = 0;

		while(day <= 2000) {
			int moveCount = 0;
			visited = new boolean[N][N];

			for (int i = 0; i < N; i++) {
				for (int j = 0; j < N; j++) {
					if (!visited[i][j]) {
						moveInfo = new ArrayList<>();
						dfs(N, L, R, i, j);

						if (moveInfo.size() > 1) {
							move();
							moveCount++;
						}
					}
				}
			}

			if (moveCount == 0) {
				break;
			}

			day++;

		}

		return day;
	}

	private static void dfs(int N, int L, int R, int x, int y) {

		visited[x][y] = true;
		moveInfo.add(new int[] {x, y});

		for (int i = 0; i < 4; i++) {
			int nextX = dx[i] + x;
			int nextY = dy[i] + y;

			if (0 <= nextX && nextX < N && 0 <= nextY && nextY < N && !visited[nextX][nextY]) {
				int result = Math.abs(country[x][y] - country[nextX][nextY]);
				if (L <= result && result <= R) {
					dfs(N, L, R, nextX, nextY);
				}
			}
		}
	}

	private static void move() {

		int human = 0;

		for(int[] pos: moveInfo) {
			human += country[pos[0]][pos[1]];
		}

		int result = human / moveInfo.size();

		for (int[] pos : moveInfo) {
			country[pos[0]][pos[1]] = result;
		}
	}

}

